### PR TITLE
add Priority monoid and PriorityAction

### PR DESCRIPTION
**Pull request by isomorphism**
As mentioned on IRC. I tried to keep things consistent with the style of other modules.

### CODE REVIEW

File: src/Data/Monoid/Priority.hs [hunk 1]
```diff
@@ -0,0 +1,93 @@
+{-# LANGUAGE MultiParamTypeClasses  #-}
+{-# LANGUAGE TypeFamilies           #-}
+{-# LANGUAGE DeriveDataTypeable     #-}
+{-# LANGUAGE DeriveFoldable         #-}
+{-# LANGUAGE DeriveFunctor          #-}
+{-# LANGUAGE DeriveTraversable      #-}
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  Data.Monoid.Priority
+-- Copyright   :  (c) 2012-2015 diagrams-core team (see LICENSE)
```
Comment (byorgey) [inline src/Data/Monoid/Priority.hs]: Surely you haven't been sitting on this code since 2012 =)

File: src/Data/Monoid/Priority.hs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+priorityAction p f = PriorityAction (Priority p (Endo f))
+
+-- | A composite @PriorityAction@ takes the least priority of its component 
+--   actions.
+instance (Ord p) => Semigroup (PriorityAction p a) where
+    PriorityAction (Priority p f) <> PriorityAction (Priority q g) 
+        = PriorityAction (Priority (min p q) (f <> g))
+
+instance (Bounded p, Ord p) => Monoid (PriorityAction p a) where
+    mempty = PriorityAction (Priority minBound mempty)
```
Comment (byorgey) [inline src/Data/Monoid/Priority.hs]: Shouldn't this be `maxBound`?

**Comment by byorgey** [comment]
Thanks!  `Priority` and its instances look good to me.  However, I don't think the `Action (PriorityAction p a) (Priority q b)` instance satisfies the `Action` laws.  For example, consider (leaving out some data constructors etc.)
```
act (PriorityAction 2 succ <> PriorityAction 1 id) (Priority 2 100)
= act (PriorityAction 1 succ) (Priority 2 100)
= Priority 2 100
```
but
```
act (PriorityAction 2 succ) (act (PriorityAction 1 id) (Priority 2 100))
= act (PriorityAction 2 succ) (Priority 2 100)
= Priority 2 101
```
Since you mentioned in IRC that `PriorityAction` was the main thing you actually wanted, I am curious to hear what your application was.

The only way I can see to make this work is to define `PriorityAction p a = [(p, Endo a)]`, where the monoid on `PriorityAction` is just the list monoid, and the action on `Priority p a` works by filtering out all the things with lower priorities and applying the composition of all the higher-priority things.

**Comment by byorgey** [comment]
And actually, even the generalized version of `PriorityAction` that I said only satisfies the laws if you restrict the functions stored in `PriorityAction` to be monoid homomorphisms (though that seems reasonable).

**Comment by isomorphism** [comment]
Oh bother. You're quite right, and while I changed the design for other reasons, now that I've thought about it more half of what I wanted to do with `PriorityAction` wouldn't have worked anyway and the other half would be better accomplished by combining `Priority` with a `Monoid`-flavored newtype wrapper specifying the operation I want.

For what it's worth, the original idea involved an abstract type where some accessors represented strictly relative values, such that they could be modified but not inspected directly. Both the relative properties and the standard ones were using `Priority` so I thought `Action` would make sense as a way to describe changes to the relative fields.

At any rate, unless inspiration strikes before dinner I'll probably just delete the `PriorityAction` stuff entirely and maybe revisit the idea later.

**Pull request closed without merging**